wallet: Track no-longer-spendable TXOs separately

achow101 opened the pull request:
In #27286, the wallet keeps track of all of its transaction outputs, even if they are already spent or are otherwise unspendable. This TXO set is iterated for balance checking and coin selection preparation, which can still be slow for wallets that have had a lot of activity. This PR aims to improve the performance of such wallets by moving UTXOs that are definitely no longer spendable to a different map in the wallet so that far fewer TXOs need to be iterated for the aforementioned functions.

Unspendable TXOs (not to be confused with Unspent TXOs) are those which have a spending transaction that has been confirmed, or are no longer valid due to reorgs. TXOs that are spent in unconfirmed transactions remain in the primary TXO set, and are filtered out of balance and coin selection as before.

achow101 commented:
@t-bast You may be interested in this for very large wallets.

t-bast commented:
Neat, thanks @achow101, @remyers will give that a try and will review the related PRs!

remyers commented:
Concept ACK - on a large synthetic test wallet I'm seeing a 250x speed up.

I created a large wallet in regtest with many fanout style transactions - 100,000 blocks each creating 1000 utxos. There are only 117 utxos currently, but the wallet.dat file is 370M.

Using the same wallet.dat file and (transaction history) `fundrawtransaction` saw a substantial speedup:
 ~5 seconds using v25.99.0-a36134fcc7b4
 ~20 milliseconds using v25.99.0-d9c4dcfe68c5 (latest [wallet-unspent-txos](https://github.com/bitcoin/bitcoin/commit/d9c4dcfe68c547d23f7687e5f0e9db363aa3b460))

I will try to recreate my large test wallet.dat file tomorrow so I can confirm the steps to reproduce it. Is there anything in particular I can do to help? Would you like more testing? code review?

NB: I had to export my original wallet descriptors and re-import them to work around the issue fixed in [#27920](https://github.com/bitcoin/bitcoin/pull/27920#pullrequestreview-1608551611). Before the export/import the wallet.dat file was 1.1G, but the speed of `fundrawtransaction` was 5 sec both in the original file and the tested one created via importing descriptors.

achow101 commented:
> Is there anything in particular I can do to help? Would you like more testing? code review?

The prerequisite PR (#27286) will need review and testing before this can be merged. That should also provide some speed up.

remyers commented:
> The prerequisite PR (#27286) will need review and testing before this can be merged. That should also provide some speed up.

Will do. I'm getting 400-500ms for the same test using just #27286, so indeed also ~100x speed-up.

murchandamus reviewed (commented):
Concept ACK

achow101 committed:
wallet: Store the TxState in each WalletTXO as well

The states will be updated whenever CWaleltTx::SetState is called too.
This is achieved by having CWalletTx::SetState take a function that
applies the new state to the specified TXO. This ensures that CWalletTx
states and WalletTXO states are kept in sync.

achow101 committed:
wallet: Get the depth in main chain for a TxState

achow101 committed:
wallet: Get the coinbase maturity state for a TXO and TxState

achow101 committed:
wallet: Use WalletTXO stored state and coinbase rather than wtx

achow101 committed:
wallet: Add m_from_me to cache "from me" status

m_from_me is used to track whether a transaction is "from me", i.e. has
any inputs which belong to the wallet.

achow101 committed:
wallet: Replace CachedTxIsFromMe with direct m_from_me lookup

Instead of looking at the cached amounts or searching every input of a
transaction each time we want to determine whether it is "from me", use
m_from_me  which stores this value for us.

achow101 committed:
wallet: Store a copy of m_from_me in WalletTXOs and use for "from me"

Since we need to know whether the transaction that creates a WalletTXO
is "from me", we should store this state in the WalletTXO too, copied
from its parent CWalletTx.

achow101 committed:
wallet: Have WalletTXOs also store parent tx time

WalletTXOs need to know their parent tx's timestamp for AvailableCoins
to work.

achow101 committed:
wallet: Make CWalletTx::m_state private with {get,set}ters

achow101 committed:
wallet: Remove unused WalletTXO::GetWalletTx()

achow101 committed:
wallet: Have IsSpent take a min_conf

A min_conf parameter is added to IsSpent so that it can set a
confirmation threshold for whether something is considered spent.

achow101 committed:
wallet: Iterate block txs in reverse on blockDisconnected

When a block is disconnected, we need to process the transactions in
reverse order so that the wallet's TXO set is updated in the correct
order.

achow101 committed:
wallet, tests: Have CreateSyncedWallet use CWallet::Create

CWallet::Create will properly connect the wallet to the chain, so we
should be doing that rather than ad-hoc chain connection.

achow101 committed:
wallet: Move definintely unusable TXOs to a separate container

Definitely unusable TXOs are those that are spent by a confirmed
transaction or were produced by a now conflicted transaction. However,
we still need them for GetDebit, so we store them in a separate
m_unusable_txos container. MarkConflicted, AbandonTransaction, and
loading (via PruneSpentTXOs) will ensure that these unusable TXOs are
properly moved.

jonatack commented:
Concept ACK, pending https://github.com/bitcoin/bitcoin/pull/27286, if this improves performance for wallets with many transactions.

achow101 commented on src/wallet/test/util.cpp:
```diff
@@ -17,18 +17,17 @@
+    std::vector<bilingual_str> warnings;
+    auto wallet = CWallet::Create(context, "", CreateMockableWalletDatabase(), WALLET_FLAG_DESCRIPTORS, error, warnings);
+
+    // Allow the fallback fee with it's default
```
Fixed

achow101 committed:
walletdb: Move ReorderTransactions to immediately after loading txs

Perform the transaction reorder upgrade immediately after loading txs
instead of waiting for the end of loading.

achow101 commented on src/wallet/transaction.h:
```diff
@@ -215,6 +215,11 @@ class CWalletTx
+     * For every isminetype, we want to track whether the transaction spends any
+     * inputs that match that isminetype.
+     */
+    std::optional<bool> m_from_me;
```
Using a default value has the potential very problematic because its value is expected to be correct. By using an optional, we can detect when there is a programming error that did not set the value.

Zeegaths commented:
## Performance Testing Results

Tested this PR against master using regtest with high-activity wallets.

### Test Setup:
- Ubuntu 22.04, CMake build
- Fresh regtest environment for both branches  
- 1,110 transactions (200 initial coinbase + 500 small sends + 100 coinbase + 300 large sends + 10 final coinbase)
- Identical test patterns on both branches

### Results:

| Operation | Master | PR Branch | Performance Change |
|-----------|--------|-----------|-------------------|
| `getwalletinfo` | 0.008s | 0.011s | +38% slower |
| `getbalance` | 0.009s | 0.010s | +11% slower |
| `listunspent` | 0.051s | 0.035s | **-31% faster** |
| `sendtoaddress` | 0.036s | 0.037s | +3% slower |

### Observations:
The TXO separation optimization shows its intended benefit in `listunspent` (31% improvement), which makes sense given that this operation iterates through unspent outputs - exactly what the PR optimizes. Other operations show minimal performance differences within expected measurement variance.

Results suggest the optimization is working as designed for the primary use case.

achow101 committed:
wallet: Include transaction version in WalletTXO

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -4448,35 +4557,108 @@ void CWallet::WriteBestBlock() const
-    for (const auto& [_, wtx] : mapWallet) {
-        RefreshTXOsFromTx(wtx);
+    for (auto& [_, wtx] : wtxOrdered) {
+        bool from_me_before = *wtx->m_from_me;
```
Can `m_from_me` be `nullopt` here ?

```suggestion
        std::optional<bool> from_me_before = wtx->m_from_me;
```

w0xlt commented:
Results with #34360 on my machine:

| Branch | ns/op | ops/s | Speedup |
| ------------- | ------------- | ------------- | ------------- |
| Master (no optimization) | 3,529,535 | 283 | baseline |
| PR #27865 (with optimization) |  6,951 |143,867 | ~508x faster |

brunoerg commented on src/wallet/wallet.cpp:
```diff
@@ -883,13 +894,14 @@ DBErrors CWallet::ReorderTransactions()
-            if (!batch.WriteTx(*pwtx))
-                return DBErrors::LOAD_FAIL;
+            if (!batch.WriteTx(*pwtx)) {
+                return false;
```
Unkilled mutant:

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index c68390c82e..3f660e2c87 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -895,7 +895,7 @@ bool CWallet::ReorderTransactions(WalletBatch& batch)

             // Since we're changing the order, write it back
             if (!batch.WriteTx(*pwtx)) {
-                return false;
+                return true;
             }
         }
     }
```

brunoerg reviewed (commented):
I ran an incremental mutation testing for `src/wallet/wallet.cpp` - it generated 58 mutants and the analysis showed a mutation score of **94.8%** (excellent). Left some comments with the unkilled mutants, feel free to address them or ignore, in any case a feedback would be valuable.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/27865/commits/52b677d4550c9d5e32e7e3cb5e9fe76eebcea7b3

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1350,6 +1393,21 @@ void CWallet::RecursiveUpdateTxState(WalletBatch* batch, const Txid& tx_hash, co
+
+                if (std::get_if<TxStateBlockConflicted>(&unusable_txo_it->second.GetState()) ||
+                    std::get_if<TxStateConfirmed>(&unusable_txo_it->second.GetState())) {
+                    continue;
```
Spent a while looking at this and I've come to the conclusion that this entire loop is actually unnecessary, so I've removed it.

w0xlt commented on src/wallet/transaction.h:
```diff
@@ -389,20 +405,35 @@ struct WalletTxOrderComparator {
+    TxState m_tx_state;
+    bool m_tx_coinbase;
+    bool m_tx_from_me;
+    int64_t m_tx_time;
```
```suggestion
    int64_t m_tx_time;
    const uint32_t m_tx_version;
```

To keep the same pattern, `m_tx_version` could be private with the getter `uint32_t GetTxVersion() const { return m_tx_version; }`.

Ant then:

```diff
diff --git a/src/wallet/spend.cpp b/src/wallet/spend.cpp
index 043f533b55..d68f441e89 100644
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -284,10 +284,10 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
                 input_bytes = CalculateMaximumSignedInputSize(txout, &wallet, &coin_control);
             }
             if (wallet.GetTxStateDepthInMainChain(txo->GetState()) == 0) {
-                if (txo->m_tx_version == TRUC_VERSION && coin_control.m_version != TRUC_VERSION) {
+                if (txo->GetTxVersion() == TRUC_VERSION && coin_control.m_version != TRUC_VERSION) {
                     return util::Error{strprintf(_("Can't spend unconfirmed version 3 pre-selected input with a version %d tx"), coin_control.m_version)};
-                } else if (coin_control.m_version == TRUC_VERSION && txo->m_tx_version != TRUC_VERSION) {
-                    return util::Error{strprintf(_("Can't spend unconfirmed version %d pre-selected input with a version 3 tx"), txo->m_tx_version)};
+                } else if (coin_control.m_version == TRUC_VERSION && txo->GetTxVersion() != TRUC_VERSION) {
+                    return util::Error{strprintf(_("Can't spend unconfirmed version %d pre-selected input with a version 3 tx"), txo->GetTxVersion())};
                 }
             }
         } else {
@@ -475,7 +475,7 @@ CoinsResult AvailableCoins(const CWallet& wallet,
 
         auto available_output_type = GetOutputType(type, is_from_p2sh);
         auto available_output = COutput(outpoint, output, nDepth, input_bytes, solvable, tx_safe, txo.GetTxTime(), tx_from_me, feerate);
-        if (txo.m_tx_version == TRUC_VERSION && nDepth == 0 && params.check_version_trucness) {
+        if (txo.GetTxVersion() == TRUC_VERSION && nDepth == 0 && params.check_version_trucness) {
             unconfirmed_truc_coins.emplace_back(available_output_type, available_output);
             auto [it, _] = truc_txid_by_value.try_emplace(outpoint.hash, 0);
             it->second += output.nValue;
```

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -4549,29 +4549,101 @@ void CWallet::RefreshTXOsFromTx(const CWalletTx& wtx)
+                assert(txos_inserted);
+            } else {
+                std::tie(txo_it, txos_inserted) = m_txos.emplace(outpoint, WalletTXO{txout, wtx.GetState(), wtx.IsCoinBase(), *wtx.m_from_me, wtx.GetTxTime(), wtx.tx->version});
+            }
```
This can be simplified.

```suggestion
            WalletTXO new_txo{txout, wtx.GetState(), wtx.IsCoinBase(), *wtx.m_from_me, wtx.GetTxTime(), wtx.tx->version};
            TXOMap& target = (m_last_block_processed_height >= 0 && IsSpent(outpoint, /*min_depth=*/1)) ? m_unusable_txos : m_txos;
            auto [txo_it, txos_inserted] = target.emplace(outpoint, std::move(new_txo));
            assert(txos_inserted);

```

w0xlt reviewed (commented):
reACK 412008bcb454b8d02f9e00f7e7a0fbb2aae5f690

w0xlt commented on src/wallet/receive.cpp:
```diff
@@ -193,32 +193,27 @@ void CachedTxGetAmounts(const CWallet& wallet, const CWalletTx& wtx,
+    const CWalletTx* wtx = wallet.GetWalletTx(txid);
+    assert(wtx);
+
+    if (!wallet.m_spend_zero_conf_change || !wtx->m_from_me) return false;
```
`m_from_me = false` passes because a populated optional converts to `true`:

```suggestion
    if (!wallet.m_spend_zero_conf_change || !wtx->m_from_me.value_or(false)) return false;
```

Example:
```c++
#include <iostream>
#include <optional>

int main()
{
    std::optional<bool> false_value{false};
    std::cout << "false_value in if: " << (static_cast<bool>(false_value) ? "true" : "false") << "\n";
}
```

Test:
```diff
diff --git a/src/wallet/receive.cpp b/src/wallet/receive.cpp
index aaf1b40955..0089da789b 100644
--- a/src/wallet/receive.cpp
+++ b/src/wallet/receive.cpp
@@ -210,7 +210,7 @@ bool CachedTxIsTrusted(const CWallet& wallet, const TxState& state, const Txid&
     const CWalletTx* wtx = wallet.GetWalletTx(txid);
     assert(wtx);
 
-    if (!wallet.m_spend_zero_conf_change || !wtx->m_from_me) return false;
+    if (!wallet.m_spend_zero_conf_change || !wtx->m_from_me.value_or(false)) return false;
 
     // Trusted if all inputs are from us and are in the mempool:
     for (const CTxIn& txin : wtx->tx->vin)
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index c5183c1c98..47ed036dab 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -90,6 +90,36 @@ BOOST_FIXTURE_TEST_CASE(update_non_range_descriptor, TestingSetup)
     }
 }
 
+BOOST_AUTO_TEST_CASE(cached_tx_is_trusted_checks_from_me_value)
+{
+    CWallet wallet(m_node.chain.get(), "", CreateMockableWalletDatabase());
+    {
+        LOCK(wallet.cs_wallet);
+        wallet.SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
+        wallet.m_spend_zero_conf_change = true;
+    }
+
+    const CKey key{GenerateRandomKey()};
+    AddKey(wallet, key);
+    const CScript script_pub_key{GetScriptForRawPubKey(key.GetPubKey())};
+
+    CMutableTransaction parent_tx;
+    parent_tx.vout.emplace_back(COIN, script_pub_key);
+    CWalletTx* parent_wtx{wallet.AddToWallet(MakeTransactionRef(parent_tx), TxStateConfirmed{GetRandHash(), /*height=*/1, /*index=*/0})};
+    BOOST_REQUIRE(parent_wtx);
+
+    CMutableTransaction child_tx;
+    child_tx.vin.emplace_back(CTxIn{parent_wtx->GetHash(), /*nIn=*/0});
+    child_tx.vout.emplace_back(COIN - DEFAULT_TRANSACTION_MAXFEE, script_pub_key);
+    CWalletTx* child_wtx{wallet.AddToWallet(MakeTransactionRef(child_tx), TxStateInMempool{})};
+    BOOST_REQUIRE(child_wtx);
+
+    LOCK(wallet.cs_wallet);
+    BOOST_CHECK(*child_wtx->m_from_me);
+    child_wtx->m_from_me = false;
+    BOOST_CHECK(!CachedTxIsTrusted(wallet, *child_wtx));
+}
+
 BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
 {
     // Cap last block file size, and mine new block in a new block file.
```

achow101 replied:
Fixed. The unit test is not meaningfully helpful.

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -4040,19 +4121,32 @@ util::Result<void> CWallet::ApplyMigrationData(WalletBatch& local_wallet_batch,
     for (const auto& [_pos, wtx] : wtxOrdered) {
+        // First update the UTXOs
+        wtx->m_txos.clear();
+        RefreshTXOsFromTx(*wtx);
```
`RefreshTXOsFromTx(*wtx)` creates or updates `WalletTXO` entries using the current `wtx->m_from_me` value. 
During migration that value may still be the legacy-wallet value, and it is recomputed later with `wtx->m_from_me = from_me`, so the refresh needs to happen after that assignment.

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 73bd712910..ab2902efea 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4121,9 +4121,7 @@ util::Result<void> CWallet::ApplyMigrationData(WalletBatch& local_wallet_batch,
         }
     }
     for (const auto& [_pos, wtx] : wtxOrdered) {
-        // First update the UTXOs
         wtx->m_txos.clear();
-        RefreshTXOsFromTx(*wtx);
         // Check it is the watchonly wallet's
         // solvable_wallet doesn't need to be checked because transactions for those scripts weren't being watched for
         bool mine = IsMine(*wtx->tx);
@@ -4131,6 +4129,7 @@ util::Result<void> CWallet::ApplyMigrationData(WalletBatch& local_wallet_batch,
         bool is_mine = mine || from_me;
         if (is_mine) {
             wtx->m_from_me = from_me;
+            RefreshTXOsFromTx(*wtx);
             // Rewrite all txs that are "mine" to ensure that any tx upgrades, including the from_me update, is written
             local_wallet_batch.WriteTx(*wtx);
         }
```

Test:

<details>

<summary>diff</summary>

```diff
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index c5183c1c98..4a18393be4 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -90,6 +90,47 @@ BOOST_FIXTURE_TEST_CASE(update_non_range_descriptor, TestingSetup)
     }
 }
 
+BOOST_AUTO_TEST_CASE(migration_refreshes_txos_after_from_me_update)
+{
+    CWallet wallet(m_node.chain.get(), "", CreateMockableWalletDatabase());
+    const CKey key{GenerateRandomKey()};
+    const CPubKey pubkey{key.GetPubKey()};
+    const CScript script_pub_key{GetScriptForDestination(PKHash{pubkey})};
+
+    LOCK(wallet.cs_wallet);
+    auto* legacy_spkm{wallet.GetOrCreateLegacyDataSPKM()};
+    BOOST_REQUIRE(legacy_spkm);
+    BOOST_REQUIRE(legacy_spkm->LoadKey(key, pubkey));
+
+    bilingual_str error;
+    auto migration_data{wallet.GetDescriptorsForLegacy(error)};
+    BOOST_REQUIRE_MESSAGE(migration_data, error.original);
+
+    CMutableTransaction tx;
+    tx.vout.emplace_back(COIN, script_pub_key);
+    CTransactionRef tx_ref{MakeTransactionRef(tx)};
+    const Txid txid{tx_ref->GetHash()};
+    auto [it, inserted]{wallet.mapWallet.emplace(std::piecewise_construct, std::forward_as_tuple(txid), std::forward_as_tuple(tx_ref, TxStateInMempool{}))};
+    BOOST_REQUIRE(inserted);
+    CWalletTx& wtx{it->second};
+    wtx.m_from_me = true;
+    wtx.nOrderPos = 0;
+    wtx.m_it_wtxOrdered = wallet.wtxOrdered.emplace(wtx.nOrderPos, &wtx);
+
+    WalletBatch batch{wallet.GetDatabase()};
+    BOOST_REQUIRE(batch.WriteBestBlock(CBlockLocator{}));
+    auto res{wallet.ApplyMigrationData(batch, *migration_data)};
+    BOOST_REQUIRE_MESSAGE(res, util::ErrorString(res).original);
+
+    const CWalletTx& migrated_wtx{wallet.mapWallet.at(txid)};
+    BOOST_REQUIRE(migrated_wtx.m_from_me.has_value());
+    BOOST_CHECK(!*migrated_wtx.m_from_me);
+
+    auto txo{wallet.GetTXO(COutPoint{txid, 0})};
+    BOOST_REQUIRE(txo);
+    BOOST_CHECK(!txo->GetTxFromMe());
+}
+
 BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
 {
     // Cap last block file size, and mine new block in a new block file.
```
</details>

achow101 replied:
Fixed, but the issue is not reachable and the unit test is not meaningfully helpful. The in memory state after migration should never be used, I'm pretty sure there are a few other places where it is out of sync. We specifically reload the wallet in order to ensure consistency before the wallet goes back into use.

w0xlt commented:
`CWallet::RemoveTxs()` deletes wallet transactions, but it was not keeping the new TXO caches (`m_txos` and `m_unusable_txos`) in sync.

Suggestion:
```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 73bd712910..a8faa6be39 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2522,6 +2522,7 @@ util::Result<void> CWallet::RemoveTxs(WalletBatch& batch, std::vector<Txid>& txs
 
     // Register callback to update the memory state only when the db txn is actually dumped to disk
     batch.RegisterTxnListener({.on_commit=[&, erased_txs]() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+        std::vector<COutPoint> spent_outpoints;
         // Update the in-memory state and notify upper layers about the removals
         for (const auto& it : erased_txs) {
             const Txid hash{it->first};
@@ -2534,14 +2535,24 @@ util::Result<void> CWallet::RemoveTxs(WalletBatch& batch, std::vector<Txid>& txs
                         break;
                     }
                 }
+                spent_outpoints.push_back(txin.prevout);
             }
             for (unsigned int i = 0; i < it->second.tx->vout.size(); ++i) {
-                m_txos.erase(COutPoint(hash, i));
+                const COutPoint outpoint{hash, i};
+                m_txos.erase(outpoint);
+                m_unusable_txos.erase(outpoint);
             }
             mapWallet.erase(it);
             NotifyTransactionChanged(hash, CT_DELETED);
         }
 
+        for (const auto& outpoint : spent_outpoints) {
+            auto txo_it = m_unusable_txos.find(outpoint);
+            if (txo_it != m_unusable_txos.end() && !std::get_if<TxStateBlockConflicted>(&txo_it->second.GetState()) && !IsSpent(outpoint, /*min_depth=*/1)) {
+                MarkTXOUsable(outpoint);
+            }
+        }
+
         MarkDirty();
     }, .on_abort={}});
```

Test:

<details>
  <summary>diff</summary>
  
```diff
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index c5183c1c98..a8ae79f51b 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -705,22 +705,52 @@ BOOST_FIXTURE_TEST_CASE(RemoveTxs, TestChain100Setup)
     m_coinbase_txns.push_back(CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey())).vtx[0]);
     auto block_tx = TestSimpleSpend(*m_coinbase_txns[0], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
     CreateAndProcessBlock({block_tx}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
+    const CTransaction block_tx_ref{block_tx};
+    auto child_tx = TestSimpleSpend(block_tx_ref, 0, key, GetScriptForRawPubKey(key.GetPubKey()));
+    CreateAndProcessBlock({child_tx}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
+    const CTransaction child_tx_ref{child_tx};
+    auto grandchild_tx = TestSimpleSpend(child_tx_ref, 0, key, GetScriptForRawPubKey(key.GetPubKey()));
+    CreateAndProcessBlock({grandchild_tx}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
 
     m_node.validation_signals->SyncWithValidationInterfaceQueue();
 
     {
-        auto block_hash = block_tx.GetHash();
+        const Txid block_hash = block_tx.GetHash();
+        const Txid child_hash = child_tx.GetHash();
+        const Txid grandchild_hash = grandchild_tx.GetHash();
         auto prev_tx = m_coinbase_txns[0];
+        const COutPoint block_outpoint{block_hash, 0};
+        const COutPoint child_outpoint{child_hash, 0};
+        const COutPoint grandchild_outpoint{grandchild_hash, 0};
 
         LOCK(wallet->cs_wallet);
         BOOST_CHECK(wallet->HasWalletSpend(prev_tx));
         BOOST_CHECK(wallet->mapWallet.contains(block_hash));
-
-        std::vector<Txid> vHashIn{ block_hash };
+        BOOST_CHECK(wallet->HasWalletSpend(MakeTransactionRef(block_tx_ref)));
+        BOOST_CHECK(wallet->mapWallet.contains(child_hash));
+        BOOST_CHECK(wallet->mapWallet.contains(grandchild_hash));
+
+        BOOST_CHECK(wallet->GetTXO(block_outpoint));
+        BOOST_CHECK(!wallet->GetTXOs().contains(block_outpoint));
+        BOOST_CHECK(wallet->GetTXO(child_outpoint));
+        BOOST_CHECK(!wallet->GetTXOs().contains(child_outpoint));
+        BOOST_CHECK(wallet->GetTXO(grandchild_outpoint));
+        BOOST_CHECK(wallet->GetTXOs().contains(grandchild_outpoint));
+
+        std::vector<Txid> vHashIn{child_hash};
         BOOST_CHECK(wallet->RemoveTxs(vHashIn));
 
-        BOOST_CHECK(!wallet->HasWalletSpend(prev_tx));
-        BOOST_CHECK(!wallet->mapWallet.contains(block_hash));
+        BOOST_CHECK(wallet->HasWalletSpend(prev_tx));
+        BOOST_CHECK(wallet->mapWallet.contains(block_hash));
+        BOOST_CHECK(!wallet->HasWalletSpend(MakeTransactionRef(block_tx_ref)));
+        BOOST_CHECK(!wallet->mapWallet.contains(child_hash));
+        BOOST_CHECK(wallet->mapWallet.contains(grandchild_hash));
+
+        BOOST_CHECK(wallet->GetTXO(block_outpoint));
+        BOOST_CHECK(wallet->GetTXOs().contains(block_outpoint));
+        BOOST_CHECK(!wallet->GetTXO(child_outpoint));
+        BOOST_CHECK(wallet->GetTXO(grandchild_outpoint));
+        BOOST_CHECK(wallet->GetTXOs().contains(grandchild_outpoint));
     }
 
     TestUnloadWallet(std::move(wallet));
```

</details>

achow101 commented:
> `CWallet::RemoveTxs()` deletes wallet transactions, but it was not keeping the new TXO caches (`m_txos` and `m_unusable_txos`) in sync.

I'm not convinced that `MarkTXOUsable` is correct or relevant. Please write a functional test that fails.

w0xlt commented:
> I'm not convinced that MarkTXOUsable is correct or relevant. Please write a functional test that fails.

Done.
The test below passes on master and on 61ba701b9e with the suggested patch, but fails on 61ba701b9e without it.

```diff
diff --git a/test/functional/wallet_importprunedfunds.py b/test/functional/wallet_importprunedfunds.py
index 95e2a5b3a4..bd08819fdc 100755
--- a/test/functional/wallet_importprunedfunds.py
+++ b/test/functional/wallet_importprunedfunds.py
@@ -157,6 +157,29 @@ class ImportPrunedFundsTest(BitcoinTestFramework):
         available_utxos = [u["txid"] for u in node.listunspent(minconf=0)]
         assert utxo["txid"] not in available_utxos, "UTXO should still be spent by conflicting tx"
 
+        self.log.info("Test removeprunedfunds restores inputs of removed confirmed spend")
+        parent_txid = node.sendtoaddress(node.getnewaddress(), 1)
+        self.generate(node, 1)
+        parent_utxo = next(
+            utxo for utxo in node.listunspent() if utxo["txid"] == parent_txid
+        )
+        parent_outpoint = (parent_txid, parent_utxo["vout"])
+
+        child_txid = node.sendall(
+            recipients=[w1.getnewaddress()], inputs=[parent_utxo]
+        )["txid"]
+        self.generate(node, 1)
+
+        def parent_is_unspent():
+            return any(
+                (utxo["txid"], utxo["vout"]) == parent_outpoint
+                for utxo in node.listunspent(minconf=0)
+            )
+
+        assert_equal(parent_is_unspent(), False)
+        node.removeprunedfunds(child_txid)
+        assert_equal(parent_is_unspent(), True)
+
 
 if __name__ == '__main__':
     ImportPrunedFundsTest(__file__).main()
```

w0xlt committed:
test: Test removedprunedfunds makes input TXOs spendable

w0xlt commented on src/wallet/test/util.cpp:
```diff
@@ -18,18 +18,17 @@
-    }
+    bilingual_str error;
+    std::vector<bilingual_str> warnings;
+    auto wallet = CWallet::CreateNew(context, "", CreateMockableWalletDatabase(), WALLET_FLAG_DESCRIPTORS, error, warnings);
```
This should fix the CI error.

```suggestion
    auto wallet = CWallet::CreateNew(context, "", CreateMockableWalletDatabase(), WALLET_FLAG_DESCRIPTORS, /*born_encrypted=*/false, error, warnings);
```

w0xlt commented:
I think the earlier `RemoveTxs()` suggestion (https://github.com/bitcoin/bitcoin/pull/27865#issuecomment-4524698348) may still be needed here. 

The unconditional `MarkTXOUsable(txin.prevout)` works when the removed tx is the only confirmed spend, but if another conflicting wallet tx is already confirmed, the same outpoint should stay unusable. 

Collecting the inputs in `spent_outpoints` lets us erase the removed txs from `mapWallet` first, then run the suggested `!IsSpent(outpoint, /*min_depth=*/1)` check against the remaining wallet transactions before moving anything back to `m_txos`.

Suggested test:

```diff
diff --git a/test/functional/wallet_importprunedfunds.py b/test/functional/wallet_importprunedfunds.py
index bd08819fdc..0b2c05c0f7 100755
--- a/test/functional/wallet_importprunedfunds.py
+++ b/test/functional/wallet_importprunedfunds.py
@@ -157,6 +157,36 @@ class ImportPrunedFundsTest(BitcoinTestFramework):
         available_utxos = [u["txid"] for u in node.listunspent(minconf=0)]
         assert utxo["txid"] not in available_utxos, "UTXO should still be spent by conflicting tx"
 
+        self.log.info("Test removeprunedfunds preserves inputs spent by another confirmed tx")
+        parent_txid = node.sendtoaddress(node.getnewaddress(), 1)
+        self.generate(node, 1)
+        parent_utxo = next(
+            utxo for utxo in node.listunspent() if utxo["txid"] == parent_txid
+        )
+        parent_outpoint = (parent_txid, parent_utxo["vout"])
+
+        child_txid = node.send(outputs=[{node.getnewaddress(): Decimal("0.5")}], inputs=[parent_utxo])["txid"]
+        child_fee = node.gettransaction(child_txid)["fee"]
+
+        # Replace child_txid with a conflicting tx and confirm it. Removing
+        # child_txid must not make parent_outpoint spendable again.
+        output_value = parent_utxo["amount"] + child_fee - Decimal("0.00001")
+        raw_conflict_tx = node.createrawtransaction(inputs=[parent_utxo], outputs=[{node.getnewaddress(): output_value}])
+        signed_conflict_tx = node.signrawtransactionwithwallet(raw_conflict_tx)
+        conflict_txid = node.sendrawtransaction(signed_conflict_tx["hex"])
+        assert_not_equal(conflict_txid, child_txid)
+        self.generate(node, 1)
+
+        def confirmed_parent_is_unspent():
+            return any(
+                (utxo["txid"], utxo["vout"]) == parent_outpoint
+                for utxo in node.listunspent(minconf=0)
+            )
+
+        assert_equal(confirmed_parent_is_unspent(), False)
+        node.removeprunedfunds(child_txid)
+        assert_equal(confirmed_parent_is_unspent(), False)
+
         self.log.info("Test removeprunedfunds restores inputs of removed confirmed spend")
         parent_txid = node.sendtoaddress(node.getnewaddress(), 1)
         self.generate(node, 1)
```

achow101 committed:
wallet: Make SubmitTxMemoryPoolAndRelay non-const

achow101 commented:
> I think the earlier `RemoveTxs()` suggestion ([#27865 (comment)](https://github.com/bitcoin/bitcoin/pull/27865#issuecomment-4524698348)) may still be needed here.

Checking the states to revert them was starting to get a bit complicated, so I ended up doing the overkill thing of recomputing all of the TXOs after removing any transaction. That should resolve any issues here.

> Suggested test:

Included the test.

***

Also ended up reworking how the tx states are set so that `CWalletTx` does not need to hold pointers to this that might disappear out from under it.

w0xlt reviewed (commented):
`AddToWallet()`'s `isInactive()` branch marks the inputs of an inactive tx usable again without checking whether another **confirmed** tx still spends them, which lets a confirmed-spent TXO back into `m_txos` and trips the new `Assert(!wallet.IsSpent(outpoint, 1))` in `GetBalance()`, aborting the node on every subsequent balance query.

Sequence:
1. UTXO `O` is spent by `T1`, which confirms → `O` moves to `m_unusable_txos`.
2. `send` with `O` as a preset input builds `T2`: `FetchSelectedInputs` resolves it via `GetTXO()` (which now also returns unusable TXOs) and has no `IsSpent` check on the preset path.
3. `CommitTransaction` calls `AddToWallet(T2, TxStateInactive{})` *before* broadcasting; the `isInactive()` branch runs `MarkTXOUsable(O)`. The broadcast then fails with `bad-txns-inputs-missingorspent`, but the wallet state was already mutated.
4. Next `getbalances`:
```
wallet/receive.cpp:256 GetBalance: Assertion `!wallet.IsSpent(outpoint, 1)' failed.
```

Suggestion:

```diff
     } else if (wtx.isInactive()) {
-        // When a transaction becomes inactive, we need to mark its inputs as usable again
+        // When a transaction becomes inactive, we need to mark its inputs as usable again,
+        // unless an input is still spent by a different confirmed transaction
         for (const CTxIn& txin : wtx.tx->vin) {
-            MarkTXOUsable(txin.prevout);
+            if (!IsSpent(txin.prevout, /*min_depth=*/1)) {
+                MarkTXOUsable(txin.prevout);
+            }
         }
     }
```

(Alternatively the guard could live inside `MarkTXOUsable()` itself, which would protect the other/future callers too.)

The test below crashes the node at `getbalances` on the current head and passes with the patch:

<details>
<summary>diff</summary>

```python
#!/usr/bin/env python3
"""Test that committing a tx that spends an already confirmed-spent input
does not resurrect the spent TXO into the usable TXO set."""

from decimal import Decimal

from test_framework.test_framework import BitcoinTestFramework
from test_framework.util import assert_equal


class WalletSpentTXOResurrectTest(BitcoinTestFramework):
    def set_test_params(self):
        self.num_nodes = 1

    def skip_test_if_missing_module(self):
        self.skip_if_no_wallet()

    def run_test(self):
        node = self.nodes[0]
        node.createwallet("resurrect")
        wallet = node.get_wallet_rpc("resurrect")
        default = node.get_wallet_rpc(self.default_wallet_name)

        self.log.info("Fund the wallet with UTXO O")
        txid = default.sendtoaddress(wallet.getnewaddress(), 1)
        self.generate(node, 1)
        utxo = next(u for u in wallet.listunspent() if u["txid"] == txid)
        outpoint = {"txid": utxo["txid"], "vout": utxo["vout"]}

        self.log.info("Spend O in T1 and confirm it -> O becomes definitely unspendable")
        t1 = wallet.send(outputs=[{default.getnewaddress(): Decimal("0.9")}],
                         inputs=[outpoint], add_inputs=False)
        assert t1["complete"]
        self.generate(node, 1)

        self.log.info("Try to double-spend O via a preset input")
        # The broadcast fails, but the tx is committed to the wallet first.
        try:
            wallet.send(outputs=[{default.getnewaddress(): Decimal("0.8")}],
                        inputs=[outpoint], add_inputs=False)
        except Exception as e:
            self.log.info(f"send reported: {e}")

        self.log.info("Balance queries must still work and O must remain spent")
        wallet.getbalances()  # aborts the node without the fix
        assert_equal(utxo["txid"] in [u["txid"] for u in wallet.listunspent()], False)


if __name__ == "__main__":
    WalletSpentTXOResurrectTest(__file__).main()
```

</details>

achow101 committed:
test: Test that send adds double spends as inactive

The send RPC can be used to create a double spend of a confirmed
transaction which is added unconditionally to the wallet as an inactive
transaction. The test ensures that the transaction is in the wallet and
that the balance is still being counted correctly.

achow101 commented:
> `AddToWallet()`'s `isInactive()` branch marks the inputs of an inactive tx usable again without checking whether another **confirmed** tx still spends them, which lets a confirmed-spent TXO back into `m_txos` and trips the new `Assert(!wallet.IsSpent(outpoint, 1))` in `GetBalance()`, aborting the node on every subsequent balance query.

I've changed `MarkTXOUsable` to always do an `IsSpent` check. Added a test in an earlier commit as well.

jeanpablojp commented on src/wallet/wallet.cpp:
```diff
@@ -4592,30 +4683,106 @@ void CWallet::RefreshTXOsFromTx(const CWalletTx& wtx)
+std::pair<CWallet::TXOMap::iterator, CWallet::TXOMap::iterator> CWallet::MaybeMarkTXOUsable(const COutPoint& outpoint)
+{
+    AssertLockHeld(cs_wallet);
+    if (IsSpent(outpoint)) {
```
`min_depth` left at the default instead of the 1 that matches `m_unusable_txos`'s definition. An output spent by a transaction that confirms and is then unconfirmed by a reorg doesn't come back, and `abandontransaction` doesn't either, since `RecursiveUpdateTxState` only walks the transaction's outputs and never its inputs. With 1 the wallet functional tests still pass and the case goes away.

jeanpablojp commented on src/wallet/wallet.cpp:
```diff
@@ -4592,30 +4683,106 @@ void CWallet::RefreshTXOsFromTx(const CWalletTx& wtx)
+                it->second.SetTxFromMe(*wtx.m_from_me);
+            } else {
+                WalletTXO txo{txout, wtx.GetState(), wtx.IsCoinBase(), *wtx.m_from_me, wtx.GetTxTime(), wtx.GetTx()->version};
+                bool is_unusable = m_last_block_processed_height >= 0 && IsSpent(outpoint, /*min_depth=*/1);
```
This is missing the `TxStateBlockConflicted` check that `PruneSpentTXOs` does. `RemoveTxs` exposes it, since it clears the containers before rebuilding. Mirroring `PruneSpentTXOs` here makes the balance match the base again.

jeanpablojp commented on src/wallet/walletdb.cpp:
```diff
@@ -1050,6 +1069,22 @@ static DBErrors LoadTxRecords(CWallet* pwallet, DatabaseBatch& batch, bool& any_
     result = std::max(result, tx_res.m_result);
 
+    // Upgrade each CWalletTx missing m_from_me
+    if (any_missing_from_me) {
```
`ReorderTransactions` used to sit after `LoadWallet`'s `if (result != DBErrors::LOAD_OK) return result;`, and now both it and the new upgrade loop run before that return. With one corrupt `tx` record, `loadwallet` fails and the intact records still get rewritten, 310 to 311 bytes here.

jeanpablojp commented on src/wallet/wallet.cpp:
```diff
@@ -4592,30 +4683,106 @@ void CWallet::RefreshTXOsFromTx(const CWalletTx& wtx)
+                it->second.SetState(wtx.GetState());
+                it->second.SetTxFromMe(*wtx.m_from_me);
+            } else {
+                WalletTXO txo{txout, wtx.GetState(), wtx.IsCoinBase(), *wtx.m_from_me, wtx.GetTxTime(), wtx.GetTx()->version};
```
`m_from_me` is an `optional<bool>` and `CWalletTx(tx, state)` doesn't set it. That's the three red fuzz jobs, all on `wallet_create_transaction`, with UBSan reporting `invalid-enum-load` and MSan the libc++ hardening assertion on `operator*`. The RefreshTXOsFromTx call there came in with #35790. `LoadToWallet` guards with `has_value()`, but the precondition isn't documented anywhere and the method is public.
